lang/java/avro/src/main/java/org/apache/avro/Conversion.java

```diff
@@ -0,0 +1,171 @@
+package org.apache.avro;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Map;
+import org.apache.avro.generic.GenericEnumSymbol;
+import org.apache.avro.generic.GenericFixed;
+import org.apache.avro.generic.IndexedRecord;
+
+/**
+ * Conversion between generic and logical type instances.
+ * <p>
+ * Instances of this class are added to GenericData to convert a logical type
+ * to a particular representation.
+ * <p>
+ * Implementations must provide:
+ * * {@link #getConvertedType()}: get the Java class used for the logical type
+ * * {@link #getLogicalTypeName()}: get the logical type this implements
+ * <p>
+ * Subclasses must also override all of the conversion methods for Avro's base
+ * types that are valid for the logical type, or else risk causing
+ * {@code UnsupportedOperationException} at runtime.
+ * <p>
+ * Optionally, use {@link #getRecommendedSchema()} to provide a Schema that
+ * will be used when a Schema is generated for the class returned by
+ * {@code getConvertedType}.
+ *
+ * @param <T> a Java type that generic data is converted to
+ */
+public abstract class Conversion<T> {
+
+  /**
+   * Return the Java class representing the logical type.
+   *
+   * @return a Java class returned by from methods and accepted by to methods
+   */
+  public abstract Class<T> getConvertedType();
+
+  /**
+   * Return the logical type this class converts.
+   *
+   * @return a String logical type name
+   */
+  public abstract String getLogicalTypeName();
+
+  public Schema getRecommendedSchema() {
+    throw new UnsupportedOperationException(
+        "No recommended schema for " + getLogicalTypeName());
+  }
+
+  public T fromBoolean(Boolean value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromBoolean is not supported for " + type.getName());
+  }
+
+  public T fromInt(Integer value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromInt is not supported for " + type.getName());
+  }
+
+  public T fromLong(Long value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromLong is not supported for " + type.getName());
+  }
+
+  public T fromFloat(Float value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromFloat is not supported for " + type.getName());
+  }
+
+  public T fromDouble(Double value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromDouble is not supported for " + type.getName());
+  }
+
+  public T fromCharSequence(CharSequence value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromCharSequence is not supported for " + type.getName());
+  }
+
+  public T fromEnumSymbol(GenericEnumSymbol value, Schema schema, LogicalType type) {
```

Comment:
This isn't an easy optimization after all, so I'm going to leave it as it is. A better optimization would be to use the same enum symbol each time rather than creating a new one.